client/gc_test.go

```diff
@@ -192,400 +188,82 @@ func TestAllocGarbageCollector_CollectAll(t *testing.T) {
-	upsertJobFn(server, job)
-
-	upsertAllocFn := func(server *nomad.Server, a *structs.Allocation) {
-		state := server.State()
-		require.NoError(state.UpsertAllocs(structs.MsgTypeTestSetup, nextIndex(), []*structs.Allocation{a}))
-	}
-
-	upsertNewAllocFn := func(server *nomad.Server, j *structs.Job) *structs.Allocation {
-		alloc := mock.Alloc()
-		alloc.Job = j
-		alloc.JobID = j.ID
-		alloc.NodeID = client.NodeID()
-
-		upsertAllocFn(server, alloc)
-
-		return alloc.Copy()
-	}
-
-	var allocations []*structs.Allocation
-
-	// Fill the node with allocations
-	for i := 0; i < maxAllocs; i++ {
-		allocations = append(allocations, upsertNewAllocFn(server, job))
+	testCases := []struct {
+		name    string
+		counter *MockAllocCounter
+		stats   *MockStatsCollector
+		expGC   bool
+	}{
+		{
+			name:    "garbage collects alloc when disk usage above threshold",
+			counter: &MockAllocCounter{},
+			stats: &MockStatsCollector{
+				availableValues: []uint64{0, 0},
+				usedPercents:    []float64{85, 85}, // above threshold
+				inodePercents:   []float64{0, 0},
+			},
+			expGC: true,
+		},
+		{
+			name:    "garbage collects alloc when inode usage above threshold",
+			counter: &MockAllocCounter{},
+			stats: &MockStatsCollector{
+				availableValues: []uint64{0, 0},
+				usedPercents:    []float64{0, 0},
+				inodePercents:   []float64{90, 90}, // above threshold
+			},
+			expGC: true,
+		},
+		{
+			name: "garbage collects alloc when liveAllocs above maxAllocs threshold",
+			counter: &MockAllocCounter{
+				allocs: 150, // above threshold
+			},
+			stats: &MockStatsCollector{
+				availableValues: []uint64{0, 0},
+				usedPercents:    []float64{0, 0},
+				inodePercents:   []float64{0, 0},
+			},
+			expGC: true,
+		},
+		{
+			name: "exits when there is no reason to GC",
+			counter: &MockAllocCounter{
+				allocs: 0,
+			},
+			stats: &MockStatsCollector{
+				availableValues: []uint64{0, 0},
+				usedPercents:    []float64{0, 0},
+				inodePercents:   []float64{0, 0},
+			},
+			expGC: false,
+		},
 	}
 
-	// Wait until the allocations are ready
-	testutil.WaitForResult(func() (bool, error) {
-		ar := len(client.getAllocRunners())
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			logger := testlog.HCLogger(t)
+			gc := NewAllocGarbageCollector(logger, tc.stats, tc.counter, gcConfig())
 
-		return ar == maxAllocs, fmt.Errorf("Expected %d allocs, got %d", maxAllocs, ar)
-	}, func(err error) {
-		t.Fatalf("Allocs did not start: %v", err)
-	})
+			// add a single alloc for garbage collection
+			ar1, cleanup1 := allocrunner.TestAllocRunnerFromAlloc(t, mock.Alloc())
+			defer cleanup1()
+			exitAllocRunner(ar1)
+			gc.MarkForCollection(ar1.Alloc().ID, ar1)
 
-	// Mark the first three as terminal
-	for i := 0; i < 3; i++ {
-		allocations[i].DesiredStatus = structs.AllocDesiredStatusStop
-		upsertAllocFn(server, allocations[i].Copy())
-	}
+			// gc
+			err := gc.keepUsageBelowThreshold()
+			must.Nil(t, err)
```

Comment:
Nitpick: this assertion has a better error message if the test fails, because it'll report the error received

```suggestion
			must.NoError(t, err)
```